internal/providers/terraform/terragrunt_hcl_provider.go

```diff
@@ -0,0 +1,426 @@
+				workingDirsToEstimate = append(workingDirsToEstimate, workingDirInfo)
+			}
+			return err
+		},
+		Parallelism: 1,
+	}
+
+	s, err := tgconfigstack.FindStackInSubfolders(terragruntOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	err = s.Run(terragruntOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	return workingDirsToEstimate, nil
+}
+
+// Downloads terraform source if necessary, then runs terraform with the given options and CLI args.
+// This will forward all the args and extra_arguments directly to Terraform.
+func (p *TerragruntHCLProvider) runTerragrunt(terragruntOptions *tgoptions.TerragruntOptions) (*terragruntWorkingDirInfo, error) {
+	terragruntConfig, err := tgconfig.ReadTerragruntConfig(terragruntOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	terragruntOptionsClone := terragruntOptions.Clone(terragruntOptions.TerragruntConfigPath)
+	terragruntOptionsClone.TerraformCommand = tgcli.CMD_TERRAGRUNT_READ_CONFIG
+
+	if terragruntConfig.Skip {
+		terragruntOptions.Logger.Infof(
+			"Skipping terragrunt module %s due to skip = true.",
+			terragruntOptions.TerragruntConfigPath,
+		)
+		return nil, nil
+	}
+
+	// We merge the OriginalIAMRoleOptions into the one from the config, because the CLI passed IAMRoleOptions has
+	// precedence.
+	terragruntOptions.IAMRoleOptions = tgoptions.MergeIAMRoleOptions(
+		terragruntConfig.GetIAMRoleOptions(),
+		terragruntOptions.OriginalIAMRoleOptions,
+	)
+
+	if err := aws_helper.AssumeRoleAndUpdateEnvIfNecessary(terragruntOptions); err != nil {
+		return nil, err
+	}
+
+	// get the default download dir
+	_, defaultDownloadDir, err := tgoptions.DefaultWorkingAndDownloadDirs(terragruntOptions.TerragruntConfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// if the download dir hasn't been changed from default, and is set in the config,
+	// then use it
+	if terragruntOptions.DownloadDir == defaultDownloadDir && terragruntConfig.DownloadDir != "" {
+		terragruntOptions.DownloadDir = terragruntConfig.DownloadDir
+	}
+
+	// Override the default value of retryable errors using the value set in the config file
+	if terragruntConfig.RetryableErrors != nil {
+		terragruntOptions.RetryableErrors = terragruntConfig.RetryableErrors
+	}
+
+	if terragruntConfig.RetryMaxAttempts != nil {
+		if *terragruntConfig.RetryMaxAttempts < 1 {
+			return nil, fmt.Errorf("Cannot have less than 1 max retry, but you specified %d", *terragruntConfig.RetryMaxAttempts)
+		}
+		terragruntOptions.RetryMaxAttempts = *terragruntConfig.RetryMaxAttempts
+	}
+
+	if terragruntConfig.RetrySleepIntervalSec != nil {
+		if *terragruntConfig.RetrySleepIntervalSec < 0 {
+			return nil, fmt.Errorf("Cannot sleep for less than 0 seconds, but you specified %d", *terragruntConfig.RetrySleepIntervalSec)
+		}
+		terragruntOptions.RetrySleepIntervalSec = time.Duration(*terragruntConfig.RetrySleepIntervalSec) * time.Second
+	}
+
+	sourceUrl, err := tgconfig.GetTerraformSourceUrl(terragruntOptions, terragruntConfig)
```

Comment:
```suggestion
	sourceURL, err := tgconfig.GetTerraformSourceURL(terragruntOptions, terragruntConfig)
```

Although if this is copied from the Terragrunt code then this suggestion is probably invalid.